Add protected data (#234)
iExecBlockchainComputing/documentation: for-developers/application-io.md

The requester uses a **dataset** to use third-party confidential data in the application.

{% hint style="info" %}

> Datasets are data in a black box, for developers seeking for enhanced interoperability see [**protected data**](#protected-data)

Reviewer: ```suggestion
Datasets are data in a black box, for developers looking for ease of use and simplicity of implementation see [**protected data**](#protected-data)
```

Reviewer: i think it could be nice to explain here for the users what is the interest for them to use a ProtectedData instead of a Dataset

Author: - I think interoperability is the key feature of protected data (create data usable by others, use others' data, discover data based on schema)
- IMO using protected data over dataset is not easier, it requires more code in the app (only JS lib to avoid this boilerplate code exists at the moment)

I prefer my proposal

Author: clarified what "interoperability" means in the context of data and apps c46843b7ff0421d57a220d072149f36d1afe11de

The **dataset** address is also exposed via `IEXEC_DATASET_ADDRESS`.

### Protected Data

The requester uses a **protected data** to use third-party confidential data in the application.

A **protected data** is a [**dataset**](#dataset) with enhanced features enabled by the [DataProtector tool](https://tools.docs.iex.ec/tools/dataProtector).

> Unlike plain datasets, **protected data** have a public schema that describes the content of the confidential data and helps data discoverability.

Reviewer 2: should be great to describe what public schema means here, where it is? or add a link if it exists somewhere else

Author: 6355da5d2266106381ac21d2e1fec492f676b152

{% code title="example protected data schema" %}

```json
{
  "name": "string",
  "phone": "i128",
  "email": "string",
  "avatar": "image/png",
  ...
}
```

{% endcode %}

#### Provisioning a protected data

> The data provider uses the [DataProtector](https://tools.docs.iex.ec/tools/dataProtector) tool to [create **protected data**](https://tools.docs.iex.ec/tools/dataProtector/dataProtectorCore/protectData) and [grant access](https://tools.docs.iex.ec/tools/dataProtector/dataProtectorCore/grantAccess) to it.

Reviewer: ```suggestion
The data provider uses the [DataProtector](https://tools.docs.iex.ec/tools/dataProtector) tool to [create **protected data**](https://tools.docs.iex.ec/tools/dataProtector/dataProtectorCore/protectData) and [grant access](https://tools.docs.iex.ec/tools/dataProtector/dataProtectorCore/grantAccess) to someone.
```

Author: The use of `it` to refer to `this protected data` was intentional, granting access involving more than `someone.` I prefer linking the grantAccess doc for more details.

> #### Consuming a protected data

Reviewer: ```suggestion
#### Consuming a protected data into an App
```

Author: the whole page use "Consuming XXX" keeping it for consistency

> Developers can use the package [`@iexec/dataprotector-deserializer`](https://www.npmjs.com/package/@iexec/dataprotector-deserializer) in there JS/TS application to extract pieces of data from a **protected data**.

Reviewer: ```suggestion
Developers can use the package [`@iexec/dataprotector-deserializer`](https://www.npmjs.com/package/@iexec/dataprotector-deserializer) in there JS/TS iexec application to extract pieces of data from a **protected data**.
```

Author: in the whole page, there are 25 occurrences of "application" all referring to "iExec application', giving the precision here and only here may be confusing
